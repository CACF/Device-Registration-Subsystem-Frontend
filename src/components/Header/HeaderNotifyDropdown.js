/*
SPDX-License-Identifier: ZLIB-ACKNOWLEDGEMENT

Copyright (c) 2018 Qualcomm Technologies, Inc.

This software is provided 'as-is', without any express or implied warranty.
In no event will the authors be held liable for any damages arising from the
use of this software.

Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

*         The origin of this software must not be misrepresented; you must
not claim that you wrote the original software. If you use this software in
a product, an acknowledgment is required by displaying the trademark/logo as
per the details provided here:
https://www.qualcomm.com/documents/dirbs-logo-and-brand-guidelines

*         Altered source versions must be plainly marked as such, and must
not be misrepresented as being the original software.

*         This notice may not be removed or altered from any source
distribution.

NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY
THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT
NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

import React, {Component} from 'react';
import {
  DropdownItem,
  DropdownMenu,
  DropdownToggle,
  Dropdown,
  Button
} from 'reactstrap';
import {
  errors,
  getAuthHeader,
  getUserInfo,
  instance,
  getStatusClass,
  getUserRole,
  getUserGroups
} from "../../utilities/helpers";
import moment from 'moment';
import i18n from './../../i18n'
import {AUTHORITY} from "../../utilities/constants";
import {last} from 'ramda'

class HeaderNotifyDropdown extends Component {

  constructor(props) {
    super(props);
    this.state = {
      dropdownOpen: false,
      notifications: [],
      loading: false,
      notificationCount: null
    };
    this.toggle = this.toggle.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.toggleDropdown = this.toggleDropdown.bind(this);
  }

  componentDidMount() {
    this.updateTokenHOC(this.checkNotifications)
  }
  updateTokenHOC = (callingFunc) => {
    let config = null;
    if(this.props.kc.isTokenExpired(0)) {
      this.props.kc.updateToken(0)
        .success(() => {
          localStorage.setItem('token', this.props.kc.token)
          config = {
            headers: getAuthHeader(this.props.kc.token)
          }
          callingFunc(config);
        })
        .error(() => this.props.kc.logout());
    } else {
      config = {
        headers: getAuthHeader()
      }
      callingFunc(config);
    }
  }
  checkNotifications = (config) => {
    const id = getUserInfo().sub
    const userGroup = getUserGroups(this.props.resources)
    if(userGroup!==""){
      instance.get(`/notification?user_id=${id}`, config)
        .then(response => {
          if (response.data.notifications.length > 0) {
            this.setState({
              notificationCount: response.data.notifications.length
            })
          }
        })
        .catch(error => {
          errors(this, error);
        })
    }
  }

  toggle(config) {
    const userGroup = getUserGroups(this.props.resources)
    if(userGroup!==""){
    this.setState({
      loading: true
    });
    const id = getUserInfo().sub
      instance.get(`/notification?user_id=${id}`, config)
        .then(response => {
          // console.log(response.data)
          this.setState({
            notifications: response.data.notifications
          }, () => {
            this.setState({
              loading: false
            });
          })
        })
        .catch(error => {
          errors(this, error);
        })
    }
  }

  handleClick(notification) {
    let data = {
      "notification_id": notification.id,
      "user_id": getUserInfo().sub
    }
    let config = {
      headers: getAuthHeader()
    }

    instance.put(`/notification`, data, config)
      .catch(error => {
        errors(this, error);
      })
  }

  getStatus(statusId) {
    if (statusId === 7) {
      return 'Rejected'
    } else if (statusId === 6) {
      return 'Approved'
    } else if(statusId === 8) {
      return 'Closed'
    } else {
      return 'Information Requested'
    }
  }
  parseMessage = (notification) => {
    let status = last(notification.message.split(' '))
    let user = getUserRole(this.props.resources)
    return `${user === AUTHORITY ? i18n.t('Notification.reviewer') : i18n.t('Notification.user')} ${notification.request_id} ${i18n.t(`Notification.${status}`)}`
  }
  toggleDropdown () {
    this.setState({
      dropdownOpen: !this.state.dropdownOpen,
    })
  }
  dropAccnt() {
    return (
      <Dropdown nav isOpen={this.state.dropdownOpen} toggle={this.toggleDropdown}>
        <DropdownToggle nav onClick={()=>this.updateTokenHOC(this.toggle)}>
          <i className="fa fa-bell-o fa-lg"></i>
          {this.state.notificationCount>0 &&
          <span className="badge badge-pill badge-danger"></span>
          }
        </DropdownToggle>
        <DropdownMenu className='dropdown-notify dropdown-menu-lg'>
          <div className="dropdown-header text-center"><b>{i18n.t('Notifications')}</b></div>
          {this.state.loading && <div className='loader-center'>
            <div className="lds-roller">
              <div></div>
              <div></div>
              <div></div>
              <div></div>
              <div></div>
              <div></div>
              <div></div>
              <div></div>
            </div>
          </div>}
          {((this.state.notifications.length > 0 && !this.state.loading) && this.state.notifications.map((notification, i) => {
            return <DropdownItem className='itenotify' href="#" key={i}>
              <Button onClick={(e) =>
                {
                  e.preventDefault()
                  this.handleClick(notification)
                }
              } className="radio-close" />
              <div className='notifymessage'>
                <p>{this.parseMessage(notification)}</p>
              </div>
              <div className='notifyfoot'>
                <p className='reqdate'>{moment(notification.generated_at).format('HH:mm DD/MM/YYYY')}</p>
                <p className='reqstatus'><span
                  className={getStatusClass(this.getStatus(notification.request_status))}>{i18n.t(this.getStatus(notification.request_status))}</span>
                </p>
              </div>
            </DropdownItem>
          })) || (!this.state.loading && <p className="no-data">{i18n.t('Notification.noNewNotifications')}</p>)}
        </DropdownMenu>
      </Dropdown>
    );
  }

  render() {
    return (
      this.dropAccnt()
    );
  }
}

export default HeaderNotifyDropdown;
