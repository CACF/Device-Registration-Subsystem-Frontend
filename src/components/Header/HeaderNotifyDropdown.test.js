/*
Copyright (c) 2018 Qualcomm Technologies, Inc.
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted (subject to the limitations in the disclaimer below) provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
* Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
* Neither the name of Qualcomm Technologies, Inc. nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

import HeaderNotifyDropdown from './HeaderNotifyDropdown'
import axios from 'axios';
import mockAxios from 'jest-mock-axios';

describe('Header Nofication Component', () => {
  afterEach(() => {
    mockAxios.reset();
  });
  test('if render correctly', () => {
    const wrapper = shallow(<HeaderNotifyDropdown/>);
    expect(wrapper).toMatchSnapshot();
  })
  test('if renders correctly again', () => {
    const wrapper = render(<HeaderNotifyDropdown/>);
    expect(wrapper).toMatchSnapshot();
  });
  test('if state renders correctly', () => {
    const wrapper = mount(<HeaderNotifyDropdown/>);
    expect(wrapper.find('HeaderNotifyDropdown').state().dropdownOpen).toBe(false)
    expect(wrapper.find('HeaderNotifyDropdown').state().notifications).toEqual([])
    expect(wrapper.find('HeaderNotifyDropdown').state().loading).toBe(false)
    expect(wrapper.find('HeaderNotifyDropdown').state().hasNotifications).toBe(false)
  });
  test('if state renders correctly', () => {
    let mockNotifications = [
      {
        "request_type": "registration_request",
        "id": 985,
        "request_status": 6,
        "message": "Your request 1579 has been approved",
        "request_id": 1579,
        "generated_at": "2018-10-29T07:03:10.824439+00:00"
      },
      {
        "request_type": "registration_request",
        "id": 94,
        "request_status": 7,
        "message": "Your request 360 has been rejected",
        "request_id": 360,
        "generated_at": "2018-10-19T07:42:03.661755+00:00"
      }
    ]
    const wrapper = mount(<HeaderNotifyDropdown/>);
    let mockNotificationResponse = {
      data: {
        notifications: mockNotifications
      },
      status: 200
    }

    mockAxios.mockResponse(mockNotificationResponse)
    wrapper.setState({
      notifications: mockNotifications,
      dropdownOpen: true,
      loading: false
    })
    wrapper.update()

    expect(wrapper.find('DropdownItem').length).toEqual(2)
    expect(wrapper.state().notifications).toEqual(mockNotifications)
    expect(wrapper.state().hasNotifications).toBe(true)
    expect(wrapper.find('Dropdown').props().isOpen).toBe(true)
    // console.log(wrapper.find('HeaderNotifyDropdown').state())
    // console.log(wrapper.find('HeaderNotifyDropdown').debug())
  });
  test('if dropdown triggers', () => {
    let mockNotifications = [
      {
        "request_type": "registration_request",
        "id": 985,
        "request_status": 6,
        "message": "Your request 1579 has been approved",
        "request_id": 1579,
        "generated_at": "2018-10-29T07:03:10.824439+00:00"
      },
      {
        "request_type": "registration_request",
        "id": 94,
        "request_status": 7,
        "message": "Your request 360 has been rejected",
        "request_id": 360,
        "generated_at": "2018-10-19T07:42:03.661755+00:00"
      }
    ]
    const wrapper = mount(<HeaderNotifyDropdown/>);
    let mockNotificationResponse = {
      data: {
        notifications: mockNotifications
      },
      status: 200
    }
    mockAxios.mockResponse(mockNotificationResponse)
    wrapper.setState({
      notifications: mockNotifications,
      dropdownOpen: true,
      loading: false
    })
    wrapper.update()

    //Toggle dropdown
    wrapper.instance().toggle()
    wrapper.update()

    mockAxios.mockResponse(mockNotificationResponse)
    wrapper.update()

    expect(wrapper.state().dropdownOpen).toBe(false)
    expect(wrapper.state().loading).toBe(false)
    expect(wrapper.find('Dropdown').props().isOpen).toBe(false)
  });
  test('if clicking on notification item render information correctly', () => {
    let mockNotifications = [
      {
        "request_type": "registration_request",
        "id": 985,
        "request_status": 6,
        "message": "Your request 1579 has been approved",
        "request_id": 1579,
        "generated_at": "2018-10-29T07:03:10.824439+00:00"
      },
      {
        "request_type": "registration_request",
        "id": 94,
        "request_status": 7,
        "message": "Your request 360 has been rejected",
        "request_id": 360,
        "generated_at": "2018-10-19T07:42:03.661755+00:00"
      }
    ]
    let spy = Sinon.spy()
    const historyMock = {push: spy};
    const wrapper = mount(<HeaderNotifyDropdown history={historyMock}/>);
    let mockNotificationResponse = {
      data: {
        notifications: mockNotifications
      },
      status: 200
    }
    mockAxios.mockResponse(mockNotificationResponse)
    wrapper.setState({
      notifications: mockNotifications,
      dropdownOpen: true,
      loading: false
    })
    wrapper.update()

    wrapper.find('DropdownItem').at(0).simulate('click')
    let notifcationclickResponse = {
      data :{
        "notifications": []
      },
      status:201
    }
    mockAxios.mockResponse(notifcationclickResponse)
    wrapper.update()

    expect(spy.calledOnce).toBe(true)
    expect(wrapper.state().hasNotifications).toBe(false)
  })
})