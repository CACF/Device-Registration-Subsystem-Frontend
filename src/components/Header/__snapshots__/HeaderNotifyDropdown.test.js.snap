// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Nofication Component if render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HeaderNotifyDropdown />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "active": false,
      "addonType": false,
      "children": Array [
        <DropdownToggle
          aria-haspopup={true}
          color="secondary"
          nav={true}
        >
          <i
            className="fa fa-bell-o fa-lg"
          />
        </DropdownToggle>,
        <DropdownMenu
          className="dropdown-notify dropdown-menu-lg"
          flip={true}
          right={true}
          tag="div"
        >
          <div
            className="dropdown-header text-center"
          >
            <b>
              Notifications
            </b>
          </div>
          <p
            className="no-data"
          >
            No new Notifications
          </p>
        </DropdownMenu>,
      ],
      "direction": "down",
      "inNavbar": false,
      "isOpen": false,
      "nav": true,
      "setActiveFromChild": false,
      "toggle": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "aria-haspopup": true,
          "children": Array [
            <i
              className="fa fa-bell-o fa-lg"
            />,
            false,
          ],
          "color": "secondary",
          "nav": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fa fa-bell-o fa-lg",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
          false,
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <div
              className="dropdown-header text-center"
            >
              <b>
                Notifications
              </b>
            </div>,
            false,
            <p
              className="no-data"
            >
              No new Notifications
            </p>,
          ],
          "className": "dropdown-notify dropdown-menu-lg",
          "flip": true,
          "right": true,
          "tag": "div",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <b>
                Notifications
              </b>,
              "className": "dropdown-header text-center",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Notifications",
              },
              "ref": null,
              "rendered": "Notifications",
              "type": "b",
            },
            "type": "div",
          },
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "No new Notifications",
              "className": "no-data",
            },
            "ref": null,
            "rendered": "No new Notifications",
            "type": "p",
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "active": false,
        "addonType": false,
        "children": Array [
          <DropdownToggle
            aria-haspopup={true}
            color="secondary"
            nav={true}
          >
            <i
              className="fa fa-bell-o fa-lg"
            />
          </DropdownToggle>,
          <DropdownMenu
            className="dropdown-notify dropdown-menu-lg"
            flip={true}
            right={true}
            tag="div"
          >
            <div
              className="dropdown-header text-center"
            >
              <b>
                Notifications
              </b>
            </div>
            <p
              className="no-data"
            >
              No new Notifications
            </p>
          </DropdownMenu>,
        ],
        "direction": "down",
        "inNavbar": false,
        "isOpen": false,
        "nav": true,
        "setActiveFromChild": false,
        "toggle": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "aria-haspopup": true,
            "children": Array [
              <i
                className="fa fa-bell-o fa-lg"
              />,
              false,
            ],
            "color": "secondary",
            "nav": true,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-bell-o fa-lg",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            false,
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <div
                className="dropdown-header text-center"
              >
                <b>
                  Notifications
                </b>
              </div>,
              false,
              <p
                className="no-data"
              >
                No new Notifications
              </p>,
            ],
            "className": "dropdown-notify dropdown-menu-lg",
            "flip": true,
            "right": true,
            "tag": "div",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <b>
                  Notifications
                </b>,
                "className": "dropdown-header text-center",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Notifications",
                },
                "ref": null,
                "rendered": "Notifications",
                "type": "b",
              },
              "type": "div",
            },
            false,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "No new Notifications",
                "className": "no-data",
              },
              "ref": null,
              "rendered": "No new Notifications",
              "type": "p",
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Header Nofication Component if renders correctly again 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "dropdown nav-item",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "aria-expanded": "false",
          "aria-haspopup": "true",
          "class": "nav-link",
          "href": "#",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "fa fa-bell-o fa-lg",
            },
            "children": Array [],
            "name": "i",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "a",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "aria-hidden": "true",
            "class": "dropdown-notify dropdown-menu-lg dropdown-menu dropdown-menu-right",
            "role": "menu",
            "tabindex": "-1",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "dropdown-header text-center",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "data": "Notifications",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "b",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "no-data",
                },
                "children": Array [
                  Object {
                    "data": "No new Notifications",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "p",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            Object {
              "attribs": Object {
                "class": "no-data",
              },
              "children": Array [
                Object {
                  "data": "No new Notifications",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "p",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "class": "dropdown-header text-center",
                },
                "children": Array [
                  Object {
                    "attribs": Object {},
                    "children": Array [
                      Object {
                        "data": "Notifications",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "b",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "aria-hidden": undefined,
            "class": undefined,
            "role": undefined,
            "tabindex": undefined,
          },
          "x-attribsPrefix": Object {
            "aria-hidden": undefined,
            "class": undefined,
            "role": undefined,
            "tabindex": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "aria-expanded": undefined,
          "aria-haspopup": undefined,
          "class": undefined,
          "href": undefined,
        },
        "x-attribsPrefix": Object {
          "aria-expanded": undefined,
          "aria-haspopup": undefined,
          "class": undefined,
          "href": undefined,
        },
      },
      Object {
        "attribs": Object {
          "aria-hidden": "true",
          "class": "dropdown-notify dropdown-menu-lg dropdown-menu dropdown-menu-right",
          "role": "menu",
          "tabindex": "-1",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "dropdown-header text-center",
            },
            "children": Array [
              Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "data": "Notifications",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "b",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "no-data",
              },
              "children": Array [
                Object {
                  "data": "No new Notifications",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "p",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "no-data",
            },
            "children": Array [
              Object {
                "data": "No new Notifications",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "p",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "dropdown-header text-center",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "data": "Notifications",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "b",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "aria-expanded": "false",
            "aria-haspopup": "true",
            "class": "nav-link",
            "href": "#",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "fa fa-bell-o fa-lg",
              },
              "children": Array [],
              "name": "i",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "a",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "aria-expanded": undefined,
            "aria-haspopup": undefined,
            "class": undefined,
            "href": undefined,
          },
          "x-attribsPrefix": Object {
            "aria-expanded": undefined,
            "aria-haspopup": undefined,
            "class": undefined,
            "href": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "aria-hidden": undefined,
          "class": undefined,
          "role": undefined,
          "tabindex": undefined,
        },
        "x-attribsPrefix": Object {
          "aria-hidden": undefined,
          "class": undefined,
          "role": undefined,
          "tabindex": undefined,
        },
      },
    ],
    "name": "li",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
